### PR TITLE
Metadata hooks, ready(), toKey() and toPath()

### PR DESCRIPTION
**Pull request by LuKks**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,9 +6,22 @@ const { FileReadStream, FileWriteStream } = require('./streams.js')
 const mutexify = require('mutexify/promise')
 
 module.exports = class Localdrive {
-  constructor (root) {
+  constructor (root, opts = {}) {
     this.root = root
+    this.metadata = opts.metadata || new Map()
```
Comment (mafintosh) [inline index.js:11]: Default should be null or noops

File: index.js [hunk 2]
```diff
@@ -6,9 +6,22 @@ const { FileReadStream, FileWriteStream } = require('./streams.js')
...
     this._lock = mutexify()
+
+    if (this.metadata instanceof Map) defaultMapToNull(this.metadata)
```
Comment (mafintosh) [inline index.js:14]: Remove this. Should just be function s

File: index.js [hunk 3]
```diff
@@ -6,9 +6,22 @@ const { FileReadStream, FileWriteStream } = require('./streams.js')
...
+  }
+
+  async ready () {}
+
+  fromPath (filename) {
+    return filename.slice(this.root.length)
```
Comment (mafintosh) [inline index.js:20]: Unixify

File: index.js (hunk 1 above)
Comment (LuKks) [reply index.js:11]: What about `{}` so all the conditions like `if (this.metadata.get)` works?
Or always functions?

File: streams.js [hunk 4]
```diff
@@ -4,13 +4,14 @@ const fsp = require('fs/promises')
 const path = require('path')
 
 class FileWriteStream extends Writable {
-  constructor (filename, lock, opts = {}) {
+  constructor (filename, drive, opts = {}) {
```
Comment (mafintosh) [inline streams.js:7]: Pass the key as an arg also

File: streams.js [hunk 5]
```diff
@@ -4,13 +4,14 @@ const fsp = require('fs/promises')
...
     super({ map })
 
-    this.executable = !!opts.executable
     this.filename = filename
+    this.drive = drive
+    this.executable = !!opts.executable
+    this.metadata = opts.metadata
```
Comment (mafintosh) [inline streams.js:13]: || null - nevet have props that are undefined, that’s the road to bugs

File: streams.js [hunk 6]
```diff
@@ -49,7 +50,16 @@ class FileWriteStream extends Writable {
 
   _destroy (cb) {
     if (!this.fd) return cb(null)
-    fs.close(this.fd, () => cb(null))
+    fs.close(this.fd, async () => {
+      if (this.metadata !== undefined) {
+        if (this.drive.metadata.set) {
+          const key = this.drive.fromPath(this.filename)
+          await this.drive.metadata.set(key, this.metadata)
```
Comment (mafintosh) [inline streams.js:57]: try catch

File: streams.js [hunk 7]
```diff
@@ -49,7 +51,19 @@ class FileWriteStream extends Writable {
...
   _destroy (cb) {
     if (!this.fd) return cb(null)
-    fs.close(this.fd, () => cb(null))
+
+    fs.close(this.fd, async () => {
+      if (this.drive.metadata.put) {
+        try {
+          await this.drive.metadata.put(this.key, this.metadata)
+        } catch (error) {
+          cb(error)
```
Comment (LuKks) [inline streams.js:60]: Is this ok?

File: test/metadata.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+  t.is((await drive.entry('/LICENSE')).value.metadata, 'Typical license')
+
+  t.ok(meta.has('/LICENSE'))
+  await drive.del('/LICENSE')
+  t.absent(meta.has('/LICENSE'))
+
+  const ws = drive.createWriteStream('/ANOTHER-LICENSE', { metadata: [1, 2, 3] })
+  await bufferToStream(Buffer.from('ISC'), ws)
+  t.alike((await drive.entry('/ANOTHER-LICENSE')).value.metadata, [1, 2, 3])
```
Comment (LuKks) [inline test/metadata.js:36]: This is returning null in MacOS, but works on Ubuntu and Windows!

**Comment by LuKks** [comment]
`toKey(filename)` is actually not needed but I maybe it's a good idea to have it?

File: index.js [hunk 9]
```diff
@@ -6,11 +6,23 @@ const { FileReadStream, FileWriteStream } = require('./streams.js')
...
 module.exports = class Localdrive {
-  constructor (root) {
-    this.root = root
+  constructor (root, opts = {}) {
+    this.root = path.resolve(root)
+    this.metadata = opts.metadata || {}
     this._lock = mutexify()
   }
 
+  async ready () {}
```
Comment (mafintosh) [inline index.js:15]: add a comment that this method is just here for compat

File: index.js [hunk 10]
```diff
@@ -6,11 +6,23 @@ const { FileReadStream, FileWriteStream } = require('./streams.js')
...
+
+  toKey (filename) {
+    if (filename.startsWith(this.root)) filename = filename.slice(this.root.length)
```
Comment (mafintosh) [inline index.js:18]: This is a bit safer I think
```js
filename = path.resolve(this.root, filename)
if (!filename.startsWith(this.root)) throw new Error('Invalid filename')
return unixPathResolve('/', filename.slice(this.root.length))
```

Comment (mafintosh) [reply index.js:20]: then relative paths also work etc

**Comment by LuKks** [comment]
@mafintosh Ready! Symlink related I will try to finally fix it in the next PR!

File: streams.js [hunk 11]
```diff
@@ -51,6 +53,17 @@ class FileWriteStream extends Writable {
...
   }
+
+  _final (cb) {
+    this._finalp().then(cb, cb)
+  }
+
+  async _finalp () {
+    const { del, put } = this.drive.metadata
+    if (this.metadata === null) {
+      if (del) await del(this.key, this.metadata)
```
Comment (mafintosh) [inline streams.js:64]: just del(this.key)

**Pull request merged**